cmd/skopeo/registry_utils.go

```diff
@@ -0,0 +1,303 @@
+	parsedURL, err := url.Parse(rawURL)
+	if err != nil {
+		return Registry{}, fmt.Errorf("Failed to parse Registry URL: %v", err)
+	}
+	baseURL := parsedURL.String()
+	parsedURL.Path = path.Join(parsedURL.Path, "v2/_catalog")
+	catalogURL := parsedURL.String()
+
+	return Registry{baseURL: baseURL, catalogURL: catalogURL}, nil
+}
+
+// Determines auth type to optimize later registry calls
+func (r *Registry) getAuthType() error {
+	resp, err := http.Get(r.catalogURL)
+	if err != nil {
+		return fmt.Errorf("Failed to connect to Catalog URL: %v", err)
+	}
+
+	headerAuth := resp.Header.Get("www-authenticate")
+	headerAuthLower := strings.ToLower(headerAuth)
+
+	if strings.Count(headerAuthLower, "bearer") > 0 {
+		fmt.Println("Using Token Auth")
+		r.authType = tokenAuth
+	} else if strings.Count(headerAuthLower, "basic") > 0 {
+		fmt.Println("Using Basic Auth")
+		r.authType = basicAuth
+	} else {
+		fmt.Println("Using No Auth")
+		r.authType = noneAuth
+	}
+	return nil
+}
+
+// Uses r.authType to abstract Authentication
+func (r *Registry) applyAuth(req *http.Request) error {
+	if r.authType == tokenAuth {
+		var token string
+		// Each catalog request returns a limit of 100 repos at a time.
+		// This conditional saves a few requests for large registry catalogs.
+		// By resetting r.catalogToken at end of getV2Catalog(), this
+		// can also be used to handle auth for getting tags.
+		// This could could be improved.
+		if len(r.catalogToken) > 0 {
+			token = r.catalogToken
+		} else {
+			// err must be declared to prevent token from being re-declared with ":=" below
+			var err error
+			token, err = getToken(req.URL.String())
+			if err != nil {
+				return err
+			}
+			r.catalogToken = token
+		}
+		req.Header.Set("Authorization", "Bearer "+token)
+	} else if r.authType == basicAuth {
+		req.SetBasicAuth(r.regUsername, r.regPassword)
+	} else {
+		// Do Nothing
+	}
+	return nil
+}
+
+// Prints list of registry repos with all available tags
+func (r *Registry) getAllReposWithTags() error {
+	err := r.getAuthType()
+	if err != nil {
+		return err
+	}
+	r.getV2Catalog()
+
+	var wg sync.WaitGroup
+	var lock sync.Mutex
+	wg.Add(len(r.repos))
+	reposWithTags := make(map[string][]string)
+
+	// Alphabetize repos
+	sort.Strings(r.repos)
+
+	for _, repo := range r.repos {
+		go func(repo string) {
+			defer wg.Done()
+			tags, _ := r.getRepoTags(repo)
+			lock.Lock()
+			reposWithTags[repo] = tags
+			lock.Unlock()
+		}(repo)
+	}
+	wg.Wait()
+
+	// Display repos with tags - format to be optimized later
+	// Consider using https://github.com/olekukonko/tablewriter
+	fmt.Println()
+	w := tabwriter.NewWriter(os.Stdout, 10, 0, 1, ' ', tabwriter.TabIndent|tabwriter.Debug)
+	fmt.Fprintln(w, "\t REPO \t TAGS \t")
+	fmt.Fprintln(w, "\t - - - - - - - - - - \t - - - - - - - - - - \t")
+	for _, repo := range r.repos {
+		tags := strings.Join(reposWithTags[repo], "  ")
+		fmt.Fprintf(w, "\t %s\t %v\t\n", repo, tags)
+	}
+	w.Flush()
```

Comment:
All of this is expected to write to `stdout` of `listOptions.run`, and it idiomatically should happen in that function or at least file; helpers can return Go data structures to be formatted by the CLI front-end.